src/architecture/unixfs.md

```diff
@@ -0,0 +1,611 @@
+      url: https://protocol.ai/
+  - name: Hugo Valtier
+    github: jorropo
+    affiliation:
+      name: Protocol Labs
+      url: https://protocol.ai/
+
+tags: ['architecture']
+order: 1
+---
+
+UnixFS is a [protocol-buffers][protobuf]-based format for describing files,
+directories and symlinks as Directed Acyclic Graphs (DAGs) in IPFS.
+
+## Nodes
+
+A :dfn[Node] is the smallest unit present in a graph, and it comes from graph
+theory. In UnixFS, there is a 1-to-1 mapping between nodes and blocks. Therefore,
+they are used interchangeably in this document.
+
+A node is addressed by a [CID]. In order to be able to read a node, its [CID] is
+required. A [CID] includes two important pieces of information:
+
+1. A [multicodec], simply known as a codec.
+2. A [multihash] used to specify the hashing algorithm, the hash parameters and
+   the hash digest.
+
+Thus, the block must be retrieved; that is, the bytes which ,when hashed using the
+hash function specified in the multihash, gives us the same multihash value back.
+
+In UnixFS, a node can be encoded using two different multicodecs, listed below. More details are provided in the following sections:
+
+- `raw` (`0x55`), which are single block :ref[Files].
+- `dag-pb` (`0x70`), which can be of any other type.
+
+## `Raw` Nodes
+
+The simplest nodes use `raw` encoding and are implicitly a :ref[File]. They can
+be recognized because their CIDs are encoded using the `raw` codec:
+
+- The file content is purely the block body.
+- They never have any children nodes, and thus are also known as single block files.
+- Their size (both `dagsize` and `blocksize`) is the length of the block body.
+
+## `dag-pb` Nodes
+
+More complex nodes use the `dag-pb` encoding. These nodes require two steps of
+decoding. The first step is to decode the outer container of the block. This is encoded using the IPLD [`dag-pb`][ipld-dag-pb] specification, which can be
+summarized as follows:
+
+```protobuf
+message PBLink {
+  // binary CID (with no multibase prefix) of the target object
+  optional bytes Hash = 1;
+
+  // UTF-8 string name
+  optional string Name = 2;
+
+  // cumulative size of target object
+  optional uint64 Tsize = 3;
+}
+
+message PBNode {
+  // refs to other objects
+  repeated PBLink Links = 2;
+
+  // opaque user data
+  optional bytes Data = 1;
+}
+```
+
+After decoding the node, we obtain a `PBNode`. This `PBNode` contains a field
+`Data` that contains the bytes that require the second decoding. These are also
+a protobuf message specified in the UnixFSV1 format:
+
+```protobuf
+message Data {
+  enum DataType {
+    Raw = 0;
+    Directory = 1;
+    File = 2;
+    Metadata = 3;
+    Symlink = 4;
+    HAMTShard = 5;
+  }
+
+  required DataType Type = 1;
+  optional bytes Data = 2;
+  optional uint64 filesize = 3;
+  repeated uint64 blocksizes = 4;
+  optional uint64 hashType = 5;
+  optional uint64 fanout = 6;
+  optional uint32 mode = 7;
+  optional UnixTime mtime = 8;
```

Comment:
We should make this compatible with proto3 which removes `optional` and `required`